### PR TITLE
Add focus trap to the RHP (v3)

### PR DESCRIPTION
**Pull request by kosmydel**
### Details

The third version of the PR as previous versions got reverted ([here](https://github.com/Expensify/App/pull/27670), and [here](https://github.com/Expensify/App/pull/24316)).

It fixed the following regressions:
https://github.com/Expensify/App/issues/31687
https://github.com/Expensify/App/issues/31725

The issue below was fixed by another PR but was also treated as a regression to this PR.
https://github.com/Expensify/App/issues/31734

### Fixed Issues

$ https://github.com/Expensify/App/issues/15631
PROPOSAL: https://github.com/Expensify/App/issues/15631#issuecomment-1641623988

### Tests

- [x] Verify that no errors appear in the JS console
- [ ] Tests from the reverted versions of the PR ([here](https://github.com/Expensify/App/pull/27670), and [here](https://github.com/Expensify/App/pull/24316))
- [x] Regression: [IOU - Blue box surrounding the currency](https://github.com/Expensify/App/issues/31687):
  - [ ] Click FAB > Request money.
  - [ ] Click on the currency.
  - [ ] Choose any currency and hit ENTER to select.
  - [ ] There will be no blue box surrounding the currency after the selection.
- [ ] Regression: [Workspace - Blue focus is shown when changing workspace name](https://github.com/Expensify/App/issues/31725):
  - [ ] Go to settings -> Workspace
  - [ ] Go to an existing workspace or create one
  - [ ] Click on the workspace name to go to the settings. (Don't go to the settings by clicking on 'Settings')
  - [ ] Change the name and Click the Enter Key on the Key Board to Save. (Don't click on the 'save' button to save.)
  - [ ] The name changes and no Blue focus is shown on the Workspace Name
- [ ] Linked regression, but not fixed in this PR: [BA - Error message doesn't stop displaying instantly after valid data is entered in date field](https://github.com/Expensify/App/issues/31734)
  - [ ] Go to Settings > Workspaces > Select Workspace > Bank Account.
  - [ ] Go trough the add bank account flow until you reach a page that you have to enter any date, like Company information or Personal Information.
  - [ ] Trigger the enter valid data error message on the date field (select the filed but do not select any date and move to another field).
  - [ ] Select a valid date.
  - [ ] The error message should disappear instantly, like the other fields do.

### QA Steps

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>
Screenshot 2023-12-21 at 15 04 36

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/104823336/faf83f9e-c3a2-41b2-8ebd-da84413482a0

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/104823336/782e0e9f-6305-44a2-9f54-e625296145bf

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/104823336/fe155066-c6d2-4e25-8224-417144cfeaf4

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/104823336/9f0b0499-51f0-4475-9f07-a45df27f530f

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/104823336/22087ead-cf8d-4d20-83af-2866e6464fef

</details>

### CODE REVIEW

**Comment by kosmydel** [comment]
I ran the tests on all platforms but Android, as it crashes on main. Will test it, when it works.

Can we do the QA tests after the C+ approves and before merging? 
Since focus trap changes affect almost every screen in the App, there is a high risk of regression. So it might be a good idea, especially since it was reverted two times. 

cc @fedirjh, @roryabraham, @luacmartins

**Comment by kosmydel** [comment]
Hey, sorry for the delay. The android was crashing, and then I had a busy week. This is ready for a review.

Can we run QA here before merging it into the main? cc @roryabraham

**Comment by kosmydel** [comment]
I've resolved conflicts and retested. 
Friendly bump @fedirjh :)

**Review by fedirjh — COMMENTED** [review]
Left minor feedback about the code.

File: src/components/ScreenWrapper.tsx
Comment (kosmydel) [reply src/components/ScreenWrapper.tsx]: Right, great idea! I'm not sure why I was fixed with that approach.

**Comment by kosmydel** [comment]
I've addressed the review and quickly tested it. cc @fedirjh

**Review by fedirjh — APPROVED** [review]
Looks good to me and tests well

**Comment by kosmydel** [comment]
Thanks @fedirjh!

Can we please do the QA before merging? Those changes got reverted two times, so it would be safer this way. cc @roryabraham

**Comment by roryabraham** [comment]
Kicked off an AdHoc build, requested regression run [here](https://expensify.slack.com/archives/C9YU7BX5M/p1705209423974619)

**Comment by kosmydel** [comment]
I've pulled the main and resolved conflicts.
cc @fedirjh for another look, to be safe

**Comment by kosmydel** [comment]
> Kicked off an AdHoc build, requested regression run [here](https://expensify.slack.com/archives/C9YU7BX5M/p1705209423974619)

Hey @roryabraham, do we have any results of the regression test?

**Comment by roryabraham** [comment]
Good question @kosmydel, sorry I didn't follow up. I didn't hear back from Applause after my initial request, so I just pinged them again for confirmation. Will let you know.

**Comment by mvtglobally** [comment]
Regression is completed. No issues

**Comment by kosmydel** [comment]
> Regression is completed. No issues

Great! Thanks!

I'm unsure if I can merge conflicts and test it this week.

**Comment by kosmydel** [comment]
I've resolved conflicts again. 

---

However, there is an issue (both on main and staging) related to navigation using the tab in the manual request flow.

Steps:
1. FAB -> Request money -> Manual
2. Type any amount, press next
3. Try using tab

---

Anyway, could you have a final look, as after the regression test I resolved conflicts three times? cc @fedirjh

**Review by fedirjh — APPROVED** [review]
Looks good. 

cc @roryabraham Let's get this merged once @kosmydel resolves the conflicts

File: src/components/FocusTrapView/index.tsx
Comment (roryabraham) [inline src/components/FocusTrapView/index.tsx:24]: Let's avoid disabling this lint rule

**Review by roryabraham — CHANGES_REQUESTED** [review]
otherwise LGTM, ready to merge after conflicts are resolved and this one comment is addressed

**Comment by kosmydel** [comment]
The ideal nav has been merged and it affected the FocusTrap behavior. The reported issue is solved. 

If we want to change anything in terms of the navigation using the Tab button, please let me know.

cc @roryabraham, @fedirjh

**Comment by roryabraham** [comment]
> The reported issue is solved.

Great, sounds like we can just close this then?

**Comment by roryabraham** [comment]
@kosmydel can we close this and create a new PR in association with https://github.com/Expensify/App/issues/36476? I also think that we have a new process for adding new external libs to the codebase that we would want to follow if we are going to proceed with adding `react-focus-trap`

**Pull request closed without merging**
